pwa/app/lib/rankingPoints.ts

```diff
@@ -0,0 +1,158 @@
+import {
+  Match,
+  MatchScoreBreakdown2016Alliance,
+  MatchScoreBreakdown2017Alliance,
+  MatchScoreBreakdown2018Alliance,
+  MatchScoreBreakdown2019Alliance,
+  MatchScoreBreakdown2020Alliance,
+  MatchScoreBreakdown2022Alliance,
+  MatchScoreBreakdown2023Alliance,
+  MatchScoreBreakdown2024Alliance,
+} from '~/api/v3';
+
+export const RANKING_POINT_LABELS: Record<number, string[]> = {
+  2016: ['Defenses Breached', 'Tower Captured'],
+  2017: ['Pressure Reached', 'All Rotors Engaged'],
+  2018: ['Auto Quest', 'Face The Boss'],
+  2019: ['Complete Rocket', 'Hab Docking'],
+  2020: ['Shield Energized', 'Shield Operational'],
+  2022: ['Cargo Bonus', 'Hangar Bonus'],
+  2023: ['Sustainability Bonus', 'Activation Bonus'],
+  2024: ['Melody Bonus', 'Ensemble Bonus'],
+};
+
+export type MatchScoreBreakdown = NonNullable<Match['score_breakdown']>['red'];
+
+export function isScoreBreakdown2016(
+  scoreBreakdown: MatchScoreBreakdown,
+): scoreBreakdown is MatchScoreBreakdown2016Alliance {
+  return (
+    (scoreBreakdown as MatchScoreBreakdown2016Alliance)
+      .teleopDefensesBreached !== undefined
```

Comment:
What's the reason for checking for specific fields rather than passing in the year? Is it to handle cases where the score breakdown is missing / incomplete?

Reply:
This uses the [type system](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) which will help significantly for `/match/` pages in the future.